File: TASMod/Console/Commands/NewGame.cs
using System;
using StardewValley;
using System.Windows.Input;
using TASMod.Recording;
using TASMod.System;

namespace TASMod.Console.Commands
{
    public class NewGame : IConsoleCommand
    {
        public override string Name => "newgame";
        public override string Description => "console menu for new save state creation";

        public string DefaultFarmerName = "TAS";
        public string DefaultFarmName = "TAS Farms";
        public string DefaultFavoriteThing = "Going Fast";
        public LocalizedContentManager.LanguageCode DefaultLanguage = LocalizedContentManager.LanguageCode.en;
        public int DefaultSeed = 0;

        public enum Stage
        {
            Farmer,
            Farm,
            Favorite,
            Seed,
            Language,
            Prefix,
            Done
        };
        public Stage CurrentStage;
        public string FarmerName;
        public string FarmName;
        public string FavoriteThing;
        public string Prefix;
        public LocalizedContentManager.LanguageCode Language;
        public int Seed;

        public override void Run(string[] tokens)
        {
            FarmerName = "";
            FarmName = "";
            FavoriteThing = "";
            Seed = 0;
            Prefix = "";
            Subscribe();

            CurrentStage = Stage.Farmer;
            Write("Enter Data for Field (empty -> default).");
            Write(MenuLine());

        }

        public string MenuLine()
        {
            switch (CurrentStage)
            {
                case Stage.Farmer:
                    return string.Format("Enter Farmer Name (default: {0}):", DefaultFarmerName);
                case Stage.Farm:
                    return string.Format("Enter Farm Name (default: {0}):", DefaultFarmName);
                case Stage.Favorite:
                    return string.Format("Enter Favorite Thing (default: {0}):", DefaultFavoriteThing);
                case Stage.Seed:
                    return string.Format("Enter Game Seed (default: {0}):", DefaultSeed);
                case Stage.Language:
                    return string.Format("Enter Language Code (options: [en,ja,ru,zh,pt,es,de,th,fr,ko,it,tr,hu]) (default: {0}):", DefaultLanguage);
                case Stage.Prefix:
                    return string.Format("Enter File Name (default: {0}_{1}):", FarmerName, Seed);
                case Stage.Done:
                    Unsubscribe();
                    Write("{0} | {1} | {2} | {3} | {4}", FarmerName, FarmName, FavoriteThing, Language, Seed);
                    CreateState();
                    return string.Format("New input created: {0}", Controller.State.Prefix);
                default:
                    return "shouldnt be here...";
            }
        }

        private void CreateState()
        {
            Controller.State = new SaveState(FarmerName, FarmName, FavoriteThing, Seed, Language);
            Controller.ResetGame = true;
            TASDateTime.setUniqueIDForThisGame((ulong)Controller.State.Seed);
            Controller.State.Prefix = Prefix;
            Controller.State.Save();
            Controller.Reset(fastAdvance: true);
        }
        public override void ReceiveInput(string input, bool writeEntry = true)
        {
            string value = input.Trim();
            switch (CurrentStage)
            {
                case Stage.Farmer:
                    CurrentStage++;
                    if (value == "")
                        FarmerName = DefaultFarmerName;
                    else
                        FarmerName = value;
                    break;
                case Stage.Farm:
                    CurrentStage++;
                    if (value == "")
                        FarmName = DefaultFarmName;
                    else
                        FarmName = value;
                    break;
                case Stage.Favorite:
                    CurrentStage++;
                    if (value == "")
                        FavoriteThing = DefaultFavoriteThing;
                    else
                        FavoriteThing = value;
                    break;
                case Stage.Seed:
                    if (value == "")
                    {
                        Seed = DefaultSeed;
                        CurrentStage++;
                    }
                    else if (Int32.TryParse(value, out Seed))
                    {
                        CurrentStage++;
                    }
                    else
                    {
                        Write("Seed {0} cannot be cast to integer type, please try again", input);
                    }
                    break;
                case Stage.Language:
                    if (value == "")
                    {
                        Language = DefaultLanguage;
                        CurrentStage++;
                    }
                    else if (Enum.TryParse(value, out LocalizedContentManager.LanguageCode lang))
                    {
                        Language = lang;
                        CurrentStage++;
                    }
                    else
                    {
                        Write("Language {0} not valid, (options: [en,ja,ru,zh,pt,es,de,th,fr,ko,it,tr,hu]) (default: {1})", value, DefaultLanguage);
                    }
                    break;
                case Stage.Prefix:
                    CurrentStage++;
                    if (value == "")
                        Prefix = string.Format("{0}_{1}", FarmerName, Seed);
                    else
                        Prefix = value;
                    break;
                default:
                    throw new Exception("shouldn't get here...");
            }
            Write(MenuLine());
        }
    }
}

File: TASMod/Assets/lua/core/scripts/newgame.lua
---@diagnostic disable: undefined-global, undefined-field
local newgame = {
    farmer_name = "abc",
    farm_name = "abc",
    favorite_thing = "abc",
    prefix = nil,
    seed = 0,
    language = "en",
}

---set the params for the new game
---@param farmer_name string @farmer name
---@param farm_name string @farm name
---@param favorite_thing string @favorite thing
---@param prefix string @save file prefix
---@param seed number @game seed value: 0-2^31-1
---@param language string @game language: en, fr, es, de, pt, ru, ja, zh
function newgame.set_params(farmer_name, farm_name, favorite_thing, prefix, seed, language)
    if farmer_name ~= nil then
        newgame.farmer_name = farmer_name
    end
    if farm_name ~= nil then
        newgame.farm_name = farm_name
    end
    if favorite_thing ~= nil then
        newgame.favorite_thing = favorite_thing
    end
    if prefix ~= nil then
        newgame.prefix = prefix
    end
    if seed ~= nil then
        newgame.seed = seed
    end

    if language ~= nil then 
        if language == "en" or language == "fr" or language == "es" or language == "de" or language == "pt" or language == "ru" or language == "ja" or language == "zh" then
            newgame.language = language
        else
            print("invalid language code, defaulting to en")
        end
    end
end

---set the farmer name
---@param farmer_name string @farmer name
function newgame.set_farmer_name(farmer_name)
    if farmer_name ~= nil then
        newgame.farmer_name = farmer_name
    end
end

---set the farm name
---@param farm_name string @farm name
function newgame.set_farm_name(farm_name)
    if farm_name ~= nil then
        newgame.farm_name = farm_name
    end
end

---set the favorite thing
---@param favorite_thing string @favorite thing
function newgame.set_favorite_thing(favorite_thing)
    if favorite_thing ~= nil then
        newgame.favorite_thing = favorite_thing
    end
end

---set the save file prefix
---@param prefix string @save file prefix
function newgame.set_prefix(prefix)
    if prefix ~= nil then
        newgame.prefix = prefix
    end
end

---set the game seed
---@param seed number @game seed value: 0-2^31-1
function newgame.set_seed(seed)
    if seed ~= nil then
        newgame.seed = seed
    end
end

---set the game language
---@param language string @game language: en, fr, es, de, pt, ru, ja, zh
function newgame.set_language(language)
    local options = {"en", "fr", "es", "de", "pt", "ru", "ja", "zh"}
    if indexof(options, language) ~= nil then
        newgame.language = language
    else
        print("invalid language code, defaulting to en")
    end
end


---generate a new game
function newgame.run()
    if newgame.prefix == nil then
        newgame.prefix = string.random(4)
        print("no prefix specified, defaulting to "..newgame.prefix)
    else
        print("prefix: "..newgame.prefix)
    end
    Controller.State.Prefix = newgame.prefix

    if newgame.seed == nil then
        newgame.seed = math.random(0, 2147483647)
        print("no seed specified, defaulting to "..newgame.seed)
    else
        print("seed: "..newgame.seed)
    end
    Controller.State.Seed = newgame.seed

    if newgame.language == nil then
        newgame.language = "en"
        print("no language specified, defaulting to "..newgame.language)
    else
        print("language: "..newgame.language)
    end
    Controller.State.Language = newgame.language

    if newgame.farmer_name == nil then
        newgame.farmer_name = "abc"
        print("no farmer name specified, defaulting to "..newgame.farmer_name)
    else
        print("farmer name: "..newgame.farmer_name)
    end
    Controller.State.FarmerName = newgame.farmer_name

    if newgame.farm_name == nil then
        newgame.farm_name = "abc"
        print("no farm name specified, defaulting to "..newgame.farm_name)
    else
        print("farm name: "..newgame.farm_name)
    end
    Controller.State.FarmName = newgame.farm_name

    if newgame.favorite_thing == nil then
        newgame.favorite_thing = "abc"
        print("no favorite thing specified, defaulting to "..newgame.favorite_thing)
    else
        print("favorite thing: "..newgame.favorite_thing)
    end
    Controller.State.FavoriteThing = newgame.favorite_thing
    advance()
    advance()
    freset(0)
end

return newgame